### PR TITLE
Connect live saving during program lifetime

### LINKED ISSUE #91: [Feature 12] Connect Live Saving
Connect saving whenever windows close and when the program terminates

### PR DESCRIPTION
**Pull request by dadler83**
New Features:
- Whenever the CommonWindowManager closes a window, the InMemoryUser is re-saved
- Window close operations have been changed

Bug Fix:
- When all windows in the program close, the InMemoryUser is saved, and the hidden windows (which is every window since we merely hide windows instead of close them) are all closed and the program ends

Testing: Test suite for saving included

### CODE REVIEW

File: src/main/java/UI/LoginUI.java [hunk 1]
```diff
@@ -37,7 +37,7 @@ public LoginUI(WindowManager existingWindows, LoginController controller) {
         this.configureFrame();
         this.centreWindow();
         this.setListeners();
-        this.setVisible(true);
+        this.programWindows.openWindow(WindowManager.LOGIN_REFERENCE_KEY);
```
Comment (MMachadoUofT) [inline src/main/java/UI/LoginUI.java:40]: This is definitely a good convention to have; that way any opening and closing of windows is always delegated to our WindowManager.

**Review by MMachadoUofT — APPROVED** [review]
This looks great! It'll certainly make testing our entities during runtime a lot easier; Oswin and I were running createCurriculum tests before, but our curriculums would be deleted. If that code was written properly, then this will allow us to load in previously-made curriculums. 
He will likely have to implement whatever convention you use to save the actual user when any given window is closed out of, unless the new WindowManager code actually accounts for that. Overall, excellent work!

**Review by Amir1441 — APPROVED** [review]
These new features will be key in the functionality of the program. We want everything to be connected and that includes the User who logs in at the start

**Review by Eggathin — APPROVED** [review]
A welcome addition that fixes the issue with the curriculums not saving after the dashboard UI is closed.

**Pull request merged**